lib/Dialect/Calyx/Transforms/CalyxLoweringUtils.cpp

```diff
@@ -744,5 +754,63 @@ BuildReturnRegs::partiallyLowerFuncToComp(mlir::func::FuncOp funcOp,
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// BuildCallInstance
+//===----------------------------------------------------------------------===//
+
+LogicalResult
+BuildCallInstance::partiallyLowerFuncToComp(mlir::func::FuncOp funcOp,
+                                            PatternRewriter &rewriter) const {
+  funcOp.walk([&](mlir::func::CallOp callOp) {
+    ComponentOp componentOp = getCallComponent(callOp);
+    SmallVector<Type, 8> resultTypes;
+    for (auto type : componentOp.getArgumentTypes())
+      resultTypes.push_back(type);
+    for (auto type : componentOp.getResultTypes())
+      resultTypes.push_back(type);
+    std::string instanceName = getInstanceName(callOp);
+
+    // Determines if an instance needs to be created. If the same function was
+    // called by CallOp before, it doesn't need to be created, if not, the
+    // instance is created.
+    if (!getState().getInstance(instanceName)) {
+      InstanceOp instanceOp =
+          createInstance(callOp.getLoc(), rewriter, getComponent(), resultTypes,
+                         instanceName, componentOp.getName());
+      getState().addInstance(instanceName, instanceOp);
+      hw::ConstantOp constantOp =
+          createConstant(callOp.getLoc(), rewriter, getComponent(), 1, 1);
+      OpBuilder::InsertionGuard g(rewriter);
+      rewriter.setInsertionPointToStart(
+          getComponent().getWiresOp().getBodyBlock());
+
+      // Creates the group that initializes the instance.
+      calyx::GroupOp groupOp = rewriter.create<calyx::GroupOp>(
+          callOp.getLoc(), "init_" + instanceName);
+      rewriter.setInsertionPointToStart(groupOp.getBodyBlock());
+      auto portInfos = instanceOp.getReferencedComponent().getPortInfo();
+      auto results = instanceOp.getResults();
+      for (const auto &[portInfo, result] : llvm::zip(portInfos, results)) {
+        if (portInfo.hasAttribute("go") || portInfo.hasAttribute("reset"))
+          rewriter.create<calyx::AssignOp>(callOp.getLoc(), result, constantOp);
+        else if (portInfo.hasAttribute("done"))
+          rewriter.create<calyx::GroupDoneOp>(callOp.getLoc(), result);
+      }
+    }
+    WalkResult::advance();
+  });
+  return success();
+}
+
+ComponentOp
+BuildCallInstance::getCallComponent(mlir::func::CallOp callOp) const {
+  std::string callee = "func_" + callOp.getCallee().str();
```

Comment:
Why use the `func_` prefix here, instead of just using the result of `getCallee()` directly?

Reply:
I set the sym_name property of func.func during the lower.
If you don't do this, it will crash the program when creating calyx.instance because `func.func` and `calyx.component` have the same `sym_name` attribute. There are two identical names in the symbol table .So the name of func.func during lower is "func_" + sym_name.If there's a better way to name it, let me know .
Sorry,my English isn't very good.

Reply:
I see, it makes sense that the sym_name attribute conflicts with the calyx.component. In that case, I think the naming makes sense, applying a `func_` prefix to the symbol of every func.func should be unique, since those symbols had to be unique initially.